orbit/pkg/table/mdm/mdm_windows.go

```diff
@@ -0,0 +1,402 @@
+
+	// getting MDM enrollment status
+	isHostMDMenrolled, enrollmentURI, err := getEnrollmentInfo()
+	if err != nil {
+		return nil, fmt.Errorf("there was a problem getting enrollment info: %s ", err)
+	}
+
+	// table returns immediately if host is not MDM enrolled
+	if isHostMDMenrolled == 0 {
+		return []map[string]string{
+			{
+				"enrollment_status":  "device_not_enrolled",
+				"enrolled_user":      "",
+				"mdm_command_input":  inputCmd,
+				"mdm_command_output": "",
+			},
+		}, nil
+	}
+
+	// Executing the input MDM command if it was present, otherwise just return with MDM enrolled info
+	if len(inputCmd) > 0 {
+
+		// performs the actual MDM cmd execution against the OS MDM stack
+		outputCmd, err := executeMDMcommand(inputCmd)
+		if err != nil {
+			return nil, fmt.Errorf("mdm command execution: %s ", err)
+		}
+
+		log.Debug().Msgf("mdm_bridge output command response:\n%s", outputCmd)
+
+		return []map[string]string{
+			{
+				"enrollment_status":  "enrolled_device",
+				"enrolled_user":      enrollmentURI,
+				"mdm_command_input":  inputCmd,
+				"mdm_command_output": outputCmd,
+			},
+		}, nil
+	} else {
+		return []map[string]string{
+			{
+				"enrollment_status": "enrolled_device",
+				"enrolled_user":     enrollmentURI,
+				"mdm_command_input": inputCmd,
+				"mdm_command_ouput": "",
+			},
+		}, nil
+	}
+}
+
+// builtin utf16tostring string expects a uint16 array but we have a pointer to a uint16
+// so we need to cast it after converting it to an unsafe pointer
+// this is a common pattern though the buffer size varies
+// see https://golang.org/pkg/unsafe/#Pointer for more details
+func localUTF16toString(ptr unsafe.Pointer) (string, error) {
+	if ptr == nil {
+		return "", errors.New("failed UTF16 conversion due to null pointer")
+	} else {
+		uint16ptrarr := (*[maxBufSize]uint16)(ptr)[:]
+		return windows.UTF16ToString(uint16ptrarr), nil
+	}
+}
+
+// getEnrollmentInfo returns the MDM enrollment status by calling into OS API IsDeviceRegisteredWithManagement()
+func getEnrollmentInfo() (uint32, string, error) {
+	// variable to hold the MDM enrollment status
+	var isDeviceRegisteredWithMDM uint32 = 0
+
+	// heap-allocated buffer to hold the URI data
+	buffUriData := make([]uint16, 0, maxBufSize)
+	if buffUriData == nil {
+		return 0, "", errors.New("failed to allocate memory for URI data")
+	}
+
+	// IsDeviceRegisteredWithManagement is going to return the MDM enrollment status
+	// https://learn.microsoft.com/en-us/windows/win32/api/mdmregistration/nf-mdmregistration-isdeviceregisteredwithmanagement
+	if returnCode, _, err := procIsDeviceRegisteredWithManagement.Call(uintptr(unsafe.Pointer(&isDeviceRegisteredWithMDM)), maxBufSize, uintptr(unsafe.Pointer(&buffUriData))); returnCode != uintptr(windows.ERROR_SUCCESS) {
+		return 0, "", fmt.Errorf("there was an error calling IsDeviceRegisteredWithManagement(): %s (0x%X)", err, returnCode)
+	} else {
+
+		uriData, err := localUTF16toString(unsafe.Pointer(&buffUriData))
```

Comment:
I don't understand why you need this call and use of `unsafe`? `buffUriData` is of type `[]uint16`, and `windows.UTF16ToString` takes a `[]uint16` as argument: https://pkg.go.dev/golang.org/x/sys/windows#UTF16ToString ? It looks like you should be able to just call `windows.UTF16ToString(buffUriData)` directly.

Reply:
Unfortunately, that does not work. The program crashes If windows.UTF16ToString(buffUriData) is used here when the device is enrolled. The main reason is that IsDeviceRegisteredWithManagement() returns a pointer to an uint16 buffer, so we need to massage the data a bit to cast it back to []uint16, which is what windows.UTF16ToString() can process. This is the main reason why localUTF16toString() is required here.

Reply:
Gotcha, thanks for the clarification!